lib/puppet/defaults.rb

```diff
@@ -1354,17 +1354,21 @@ def self.initialize_default_settings!(settings)
       executable.",
     },
     :serverport => {
-      :default    => "8140",
+      :default    => 8140,
       :desc       => "The default port puppet subcommands use to communicate
       with Puppet Server. (eg `puppet facts upload`, `puppet agent`). May be
       overridden by more specific settings (see `ca_port`, `report_port`).",
     },
     :masterport => {
       :default    => "$serverport",
+      :type       => :port,
       :desc       => "The default port puppet subcommands use to communicate
       with Puppet Server. (eg `puppet facts upload`, `puppet agent`). May be
       overridden by more specific settings (see `ca_port`, `report_port`).",
-      :hook => proc { |value| Puppet[:serverport] = value }
+      :hook => proc do |value|
+        munge(value)
+        Puppet[:serverport] = value
```

Comment:
I think we want to assign the munged value to `Puppet[:serverport]`, otherwise the latter could be set to a string if called programmatically:

```ruby
require 'puppet'
Puppet.initialize_settings
Puppet[:masterport] = "88"
puts "masterport: " + Puppet[:masterport].inspect
puts "serverport: " + Puppet[:serverport].inspect
```

generates

```
$ bx ruby setting.rb
masterport: 88
serverport: "88"
```

Also wanted to mention that settings does do some magic coersion/translation in https://github.com/puppetlabs/puppet/blob/main/lib/puppet/settings/value_translator.rb#L8. I don't think it affects this PR since we want to restrict it to a valid port, but the translator is something to watch out, for example https://tickets.puppetlabs.com/browse/PUP-8289?focusedCommentId=519858&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-519858

Reply:
> I think we want to assign the munged value to `Puppet[:serverport]`, otherwise the latter could be set to a string if called programmatically

Yep. This should also be fixed if we're setting the `:port` type on the `:serverport` setting.

> Also wanted to mention that settings does do some magic coersion/translation in https://github.com/puppetlabs/puppet/blob/main/lib/puppet/settings/value_translator.rb#L8.

Hmm. This seems interesting, from what I see it's used only when parsing the config file, and when parsing CLI arguments.

Reply:
I think this needs a rebase do to https://github.com/puppetlabs/puppet/blob/729ad1867faa278a398bf45db95cbedf63d32e10/lib/puppet/defaults.rb#L1384-L1386

Reply:
Yeah this is still WIP, I haven't pushed the latest changes yet